Cargo.toml

```diff
@@ -10,8 +10,7 @@ readme = "README.md"
 description = "Inlines modules in Rust source code for source analysis"
 
 [dependencies]
-syn = { version = "^1.0.0", default-features = false, features = ["parsing", "printing", "full", "visit-mut"] }
+# We support syn 1.x and 2.x
+syn = { version = ">=1.0.0, <3.0.0", default-features = false, features = ["parsing", "printing", "full", "visit-mut"] }
```

Comment:
@Manishearth please do not do this in any other crates.

This does not mean "syn-inline-mod works with both syn 1 and syn 2". It means "syn-inline-mod works with a random one of syn 1 or syn 2 that you can't control. But only one of them."

So for example if someone downstream has:

```toml
[dependencies]
serde = "1"
syn = "1"
syn-inline-mod = "0.5.0"
```

suppose Cargo resolves the random choice to be that syn-inline-mod depends on syn 1 because syn 2 isn't already in the workspace. Their code relies on using syn 1 types with syn-inline-mod's public API.

Then they do `cargo update` and serde gets updated to syn 2 in a non-breaking manner (syn is not in serde's public API) and syn-inline-mod gets updated to 0.5.1. Now Cargo resolves the random choice to be that syn-inline-mod only works with syn 2 and you've broken this person's code.

Reply:
I'm not planning on doing this elsewhere. I understand the Cargo behavior here; I disagree that you can't control it, it's just a bit more work.


I opened this PR to give the maintainer the choice, especially if they wanted to do a major version bump. It's more flexible for the user to allow them to pick and choose how the dependency gets resolved; and in the case of a major version bump the problem you sketch isn't a problem. It will still be a problem for new users who want to use the crate, but it is possible to control how this gets resolved.

I do agree there is a tradeoff here.

Reply:
> I understand the Cargo behavior here; I disagree that you can't control it, it's just a bit more work.

> it is possible to control how this gets resolved.

I guess this is going to be a learning experience for me. How do I control it?

For example this is the user's library:

```toml
[dependencies]
serde_derive = "1.0.160"
syn = "1"
syn-inline-mod = { git = "https://github.com/TedDriggs/syn-inline-mod", rev = "refs/pull/21/head" }
```

```rust
pub fn repro(error: syn_inline_mod::Error) -> syn::Error {
    match error {
        syn_inline_mod::Error::Parse(error) => error,
        syn_inline_mod::Error::Io(_) => todo!(),
    }
}
```

How do they make this compile syn-inline-mod against syn 1 such that their code compiled?

Reply:
```
[13:33:35] मanishearth@manishearth-glaptop2 ~/sand/crates/syntest ^_^ 
$ rm Cargo.lock 
[13:33:38] मanishearth@manishearth-glaptop2 ~/sand/crates/syntest ^_^ 
$ cargo c
    Updating crates.io index
    Updating git repository `https://github.com/TedDriggs/syn-inline-mod`
    Checking syntest v0.1.0 (/home/manishearth/sand/crates/syntest)
error[E0308]: mismatched types
<snip>
error: could not compile `syntest` due to previous error
[13:33:42] मanishearth@manishearth-glaptop2 ~/sand/crates/syntest O_o 
$ cargo update --aggressive -p syn-inline-mod
    Updating git repository `https://github.com/TedDriggs/syn-inline-mod`
    Updating crates.io index
[13:33:49] मanishearth@manishearth-glaptop2 ~/sand/crates/syntest ^_^ 
$ cargo c
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
```

Reply:
basically, `cargo update --aggressive -p syn-inline-mod` is pretty awesome

Reply:
it would be nice if `--aggressive` could take a crate spec argument so you can select a _specific_ dep to aggressively update (and allow using it with `--precise`) but what it does so far is often decent enough.

There's also just manually editing lockfiles but I do try really really hard to never have to do that. This is a change that is possible to make by a relatively simple manual edit since you don't have to touch any of the hashes.